File: Q17 spiral matrix.py
class Solution:
    def spiralOrder(self, matrix) :
        ans = []
        m = len(matrix)
        n = len(matrix[0])
        if m == 0:
            return ans
        total = m * n

        starting_row = 0
        ending_col = n - 1
        ending_row = m - 1
        starting_col = 0
        count = 0
        while count < total:
            # starting row
            for i in range(starting_col, ending_col + 1):
                if count < total:
                    ans.append(matrix[starting_row][i])
                    count += 1
            starting_row += 1
            # ending column
            for i in range(starting_row, ending_row + 1):
                if count < total:
                    ans.append(matrix[i][ending_col])
                    count += 1
            ending_col -= 1
            # ending row
            for i in range(ending_col, starting_col - 1, -1):
                if count < total:
                    ans.append(matrix[ending_row][i])
                    count += 1
            ending_row -= 1
            # starting column
            for i in range(ending_row, starting_row - 1, -1):
                if count < total:
                    ans.append(matrix[i][starting_col])
                    count += 1
            starting_col += 1
        return ans

File: Q17 spiral matrix.js
var spiralOrder = function(matrix) {
    let ans = [];
    let m = matrix.length;
    let n = matrix[0].length;
    if (m === 0) return ans;
    let total = m * n;

    let startingRow = 0;
    let endingCol = n - 1;
    let endingRow = m - 1;
    let startingCol = 0;
    let count = 0;
    while (count < total) {
        // starting row
        for (let i = startingCol; i <= endingCol && count < total; i++) {
            ans.push(matrix[startingRow][i]);
            count++;
        }
        startingRow++;
        // ending column
        for (let i = startingRow; i <= endingRow && count < total; i++) {
            ans.push(matrix[i][endingCol]);
            count++;
        }
        endingCol--;
        // ending row
        for (let i = endingCol; i >= startingCol && count < total; i--) {
            ans.push(matrix[endingRow][i]);
            count++;
        }
        endingRow--;
        // starting column
        for (let i = endingRow; i >= startingRow && count < total; i--) {
            ans.push(matrix[i][startingCol]);
            count++;
        }
        startingCol++;
    }
    return ans;
};

